coding/fuzz/src/lib.rs

```diff
@@ -1,61 +1,67 @@
 use arbitrary::{Arbitrary, Unstructured};
 use commonware_coding::{Config, Scheme};
+use std::iter;
+
+#[derive(Debug)]
+struct Shuffle {
+    choices: Vec<usize>,
+}
+
+impl Shuffle {
+    fn arbitrary<'a>(u: &mut Unstructured<'a>, len: usize) -> arbitrary::Result<Self> {
+        let choices = (0..len - 1)
+            .map(|i| Ok(i + u.choose_index(len - i)?))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        Ok(Self { choices })
+    }
+
+    fn shuffle<T>(&self, data: &mut [T]) {
+        assert_eq!(
+            self.choices.len() + 1,
+            data.len(),
+            "data length must match shuffle length"
+        );
+        for (i, &choice) in self.choices.iter().enumerate() {
+            data.swap(i, choice);
+        }
+    }
+}
 
 #[derive(Debug)]
 pub struct FuzzInput {
     min: u16,
     recovery: u16,
     to_use: u16,
     data: Vec<u8>,
-    shuffle_bytes: Vec<u8>,
+    shuffle: Shuffle,
 }
 
 impl<'a> Arbitrary<'a> for FuzzInput {
     fn arbitrary(u: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
-        // We need to generate parameters which satisfy the conditions of valid RS coding,
-        // which are such that if min <= 2^16 - 2^n, then recovery <= 2^n.
-        let n: u64 = u.int_in_range(0..=15)?;
-        let mut min = u.int_in_range(1..=u16::try_from((1 << 16) - (1 << n)).unwrap())?;
-        let mut recovery = u.int_in_range(1..=u16::try_from(1 << n).unwrap())?;
-        // Correction to make sure that we can fit min + recovery in a u16.
-        if min.checked_add(recovery).is_none() {
-            if recovery > 1 {
-                recovery -= 1;
-            } else {
-                min -= 1;
-            }
-        }
+        let min = u.int_in_range(1..=((1 << 10) - 1))?;
+        let recovery = u.int_in_range(min..=((1 << 10) - 1))?;
         let to_use = u.int_in_range(min..=min + recovery)?;
-        let data_len = u.int_in_range(0..=u32::MAX)?; // data.len() <= u32:Max
-        let data = u.bytes(data_len as usize)?.to_vec();
-        let shuffle_bytes = u.bytes(8)?.to_vec();
+        let data = u.arbitrary::<Vec<u8>>()?;
```

Comment:
I suspect we'll save a good bit of memory if we better cap this (unless I'm missing something)?

Reply:
Not sure why this worked before tbh...

Reply:
The issue is more so how the other parameters affect the number of samples you need, and the minimum size of each shard, versus the data itself. The way the fuzzing suite works is that the fuzzer won't just generate a big vector that exceeds the memory budget, but it's not able to peer into the code to avoid doing things with consequences later on.